src/org/sosy_lab/common/collect/SkipList.java

```diff
@@ -163,37 +166,86 @@ private void readObject(ObjectInputStream pIn) throws IOException, ClassNotFound
   private static final double ONE_HALF_LOG = Math.log(0.5);
   private Random randomGenerator = new Random();
 
-  private final Comparator<? super T> comparator;
+  private final @Nullable Comparator<? super T> comparator;
   private Node<T> head = createHead();
   private Node<T> tail = head;
   private int size = 0;
 
-  @SuppressWarnings("unchecked")
-  public SkipList(@Nullable Comparator<? super T> pComparator) {
-    if (pComparator == null) {
-      comparator = (Comparator<? super T>) Comparator.naturalOrder();
-    } else {
-      comparator = pComparator;
-    }
+  private SkipList(Comparator<? super T> pComparator) {
+    Preconditions.checkNotNull(pComparator);
+    comparator = pComparator;
   }
 
-  public SkipList(Collection<? extends T> pCollection) {
+  private SkipList(Iterable<? extends T> pCollection) {
     this();
-    boolean changed = addAll(pCollection);
-    assert changed;
+    addAll0(pCollection);
   }
 
-  public SkipList(SortedSet<T> pSortedSet) {
-    this(pSortedSet.comparator());
-    boolean changed = addAll(pSortedSet);
-    assert changed;
+  private SkipList(Iterable<? extends T> pCollection, Comparator<? super T> pComparator) {
+    this(pComparator);
+    addAll0(pCollection);
   }
 
-  @SuppressWarnings("unchecked")
-  public SkipList() {
-    comparator = (Comparator<? super T>) Comparator.naturalOrder();
+  private SkipList() {
+    comparator = null;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1) {
+    SkipList<T> list = new SkipList<>();
+    list.add(p1);
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2));
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2, T p3) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2, p3));
+    return list;
   }
 
+  @SafeVarargs
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2, T p3, T... pOtherElements) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2, p3));
+    for (T v : pOtherElements) {
+      list.add(v);
+    }
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> create() {
```

Comment:
Why not name this `of` in the Guava style?

Reply:
I used the [Guava Wiki](https://github.com/google/guava/wiki/CollectionUtilitiesExplained#static-constructors) as reference,
which states that newer Collection classes use `create()`.

I also find it more speaking than an `of` without parameters, and it can be easily found.

Reply:
This wiki page uses `create()` only as an example and does not say that newer collection classes use `create()`. However, mutable collections might be a better place to look at for reference than immutable collections if ours is mutable. This would indicate to have `create()` and no other `of` methods.

Reply:
Personally, I don't have any use case where you would need a sorted set of only `n` elements,
with `n` being that small that you write it down by hand (except for testing).
Thus, I don't require the `of` methods, but maybe you have a good use case?

Reply:
Then let's remove it.